packages/interaction-tracking/src/InteractionTracking.js

```diff
@@ -0,0 +1,332 @@
+  ((interactionsRef: any): InteractionsRef).current = new Set();
+
+  try {
+    return callback();
+  } finally {
+    ((interactionsRef: any): InteractionsRef).current = prevInteractions;
+  }
+}
+
+export function getCurrent(): Set<Interaction> | null {
+  if (!enableInteractionTracking) {
+    return null;
+  } else {
+    return ((interactionsRef: any): InteractionsRef).current;
+  }
+}
+
+export function getThreadID(): number {
+  return ++threadIDCounter;
+}
+
+export function subscribe(subscriber: Subscriber): void {
+  if (enableInteractionTracking && enableInteractionTrackingObserver) {
+    invariant(
+      ((subscriberRef: any): SubscriberRef).current === null ||
+        ((subscriberRef: any): SubscriberRef).current === subscriber,
+      'Only one interactions subscriber may be registered at a time.',
+    );
+    ((subscriberRef: any): SubscriberRef).current = subscriber;
+  }
+}
+
+export function track(
+  name: string,
+  callback: Function,
+  threadID: number = DEFAULT_THREAD_ID,
+): any {
+  if (!enableInteractionTracking) {
+    return callback();
+  }
+
+  const interaction: Interaction = {
+    id: interactionIDCounter++,
+    name,
+    timestamp: now(),
+  };
+
+  const prevInteractions = ((interactionsRef: any): InteractionsRef).current;
+
+  // Tracked interactions should stack/accumulate.
+  // To do that, clone the current interactions.
+  // The previous set will be restored upon completion.
+  const interactions = new Set(prevInteractions);
+  interactions.add(interaction);
+  ((interactionsRef: any): InteractionsRef).current = interactions;
+
+  if (enableInteractionTrackingObserver) {
+    // Update before calling callback in case it schedules follow-up work.
+    interaction.__count = 1;
+
+    let caughtError = null;
+    let returnValue;
+    const subscriber = ((subscriberRef: any): SubscriberRef).current;
+
+    if (subscriber !== null) {
+      try {
+        subscriber.onInteractionTracked(interaction);
+        subscriber.onWorkStarted(interactions, threadID);
+      } catch (error) {
+        caughtError = caughtError || error;
+      }
+    }
+
+    try {
+      returnValue = callback();
+    } catch (error) {
+      caughtError = caughtError || error;
+    }
+
+    ((interactionsRef: any): InteractionsRef).current = prevInteractions;
+
+    if (subscriber !== null) {
+      try {
+        subscriber.onWorkStopped(interactions, threadID);
+      } catch (error) {
+        caughtError = caughtError || error;
+      }
+    }
+
+    interaction.__count = ((interaction.__count: any): number) - 1;
+
+    // If no async work was scheduled for this interaction,
+    // Notify subscribers that it's completed.
+    if (subscriber !== null && ((interaction.__count: any): number) === 0) {
+      try {
+        subscriber.onInteractionScheduledWorkCompleted(interaction);
+      } catch (error) {
+        caughtError = caughtError || error;
+      }
+    }
+
+    if (caughtError !== null) {
```

Comment:
Nit: this would swallow `throw null`.

In React, we solve this by having a wrapper function that toggles a boolean if an error was set.

Reply:
True. That occurred to me but seemed unlikely. I'll replace it with a bool though!